storage/session.go

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (C) 2023 Nuts community
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package storage
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/nuts-foundation/nuts-node/storage/log"
+	"strings"
+	"sync"
+	"time"
+)
+
+var _ SessionDatabase = (*inMemorySessionDatabase)(nil)
+var _ SessionStore = (*inMemorySessionStore)(nil)
+
+var sessionStorePruneInterval = 10 * time.Minute
+
+type expiringEntry struct {
+	// Value stores the actual value as JSON
+	Value  string
+	Expiry time.Time
+}
+
+// SessionDatabase is an in memory database that holds session data on a KV basis.
+// Keys could be access tokens, nonce's, authorization codes, etc.
+// All entries are stored with a TTL, so they will be removed automatically.
+type inMemorySessionDatabase struct {
```

Comment:
I agree in general; who's going to use this? It feels to detached from the current storage functionality and types; (`Provider` providing `KVStore`s, and `SessionDatabase` providing `SessionStore`s).

I think you could even build this on top of `KVStore` (and use in-memory BBolt db indeed)?

Reply:
It must be detached from the current storage since it has a totally different goal. Session storage must never be written to disk and must be clustered (real time consistency, not eventual through copying). Therefore there's no logic in using bbolt or the KVStore abstraction.

Reply:
The storage engine isn't exclusively for KVStore, any SQL client could be placed in storage as well.

Reply:
Although session storage is often persistable (to handle continuance after crash/reboot/update), we probably don't need it (at least not at this point). The simplest solution is also the easiest to migrate later on, so I'm fine with a separate in-memory store, as long as the interface stays simple.